### ISSUE TITLE
Разноцветные модели

### ISSUE THREAD

**Issue by oleg-medovikov** [mirmik/zencad#39]
Добрый день.
Главный вопрос такой: как сделать модель состоящую из элементов разных цветов, чтобы ее как функцию потом можно было вставить в другой проект и при этом перемещать и вращать до нужной позиции.  

у меня пока получается только так, при этом работает перемещение, но вращение уже не получается реализовать, все разлетается в стороны.  
```
def element1():
     return m
def element2():
    return m

def model(x,y,z):
    disp(element1().translate(x,y,z),color())
    disp(element2().translate(x,y,z),color())


model(1,2,3)
```

**Comment by mirmik** [comment]
На текущий момент интерактивный объект может быть покрашен только в один цвет. Ядро, насколько я смог понять, позволяет многоцветные и текстурированные формы, но сейчас этот функционал в zencad не реализован.

Для того, чтобы оперировать с одновременно интерактивными объектами, можно использовать механизм сборок:

Пример:

```
#!/usr/bin/env python3
#coding: utf-8

from zencad import *
import zencad.assemble

a=box(10)
b=box(10).moveX(10)

u = zencad.assemble.unit()
u.add_shape(a, color.red)
u.add_shape(b, color.blue)

u.relocate(moveZ(10) * rotateZ(deg(45)))

disp(u)
show()
```

**Comment by oleg-medovikov** [comment]
Именно то, что я и хотел. 
Спасибо.

**Comment by oleg-medovikov** [comment]
https://github.com/oleg-medovikov/zencad

вот результат, если интересно. 
это arduino + CNC shield v3
с достаточно точными размерами

**Comment by mirmik** [comment]
Могу я скриншоты этой и другой вашей модели (которая дельта робот) использовать как иллюстрации в будущей статье про zencad?

**Comment by oleg-medovikov** [comment]
да, только те картинки устарели, я переделываю модель так, чтобы ее можно было печатать на малом столе 200х200 мм и добавляю самое разное. Держатель для концевиков, отверстия для проводов, место под камеру и так далее.   
  
![screenshot-2020-02-22-08:47:18](https://user-images.githubusercontent.com/57829409/75087375-66489980-5550-11ea-92ad-09eaf44d43ef.png)

**Comment by oleg-medovikov** [comment]
Долго добивался, чтобы при изменении радиуса базы, оно правильно раздвигалось.
![screenshot-2020-02-02-15:32:16](https://user-images.githubusercontent.com/57829409/75087508-5631b980-5552-11ea-873e-cbc799551bdd.png)
![screenshot-2020-02-02-15:32:40](https://user-images.githubusercontent.com/57829409/75087521-7ceff000-5552-11ea-91af-1ab99ac50830.png)
![screenshot-2020-02-02-15:55:06](https://user-images.githubusercontent.com/57829409/75087519-73668800-5552-11ea-9a89-a85cefbceab2.png)





 В жизни оно вот так выглядит 
![BfEa0KfrWqc](https://user-images.githubusercontent.com/57829409/75087479-f4714f80-5551-11ea-84bb-d107801ff054.jpg)
![sv5M9dJQzrk](https://user-images.githubusercontent.com/57829409/75087480-f804d680-5551-11ea-8a5f-fec8c1dae450.jpg)

**Comment by mirmik** [comment]
Капитальненько...

**Comment by oleg-medovikov** [comment]
Рано закрыл тему. 
Вот такая проблема:
Есть у меня assemble.unit и я хочу несколько раз использовать его в сцене. 
Я планировал делать 
```
relocate(1 место)
disp()
relocate(2 место)
disp()
...
```
Но как выяснилось, отображается только последнее положение, что меня удивило.
Тогда попробовал вот так:

```
def extrans(): return multitransform([
    translate(0,-rad-45,27) * rotateZ(deg(-90)) * rotateY(deg(165)),
    translate(0,-rad,-70),
    nulltrans()
])

disp(extrans(konchevik))
```
Выводит ошибку:

`TypeError: extrans() takes 0 positional arguments but 1 was given`

**Comment by mirmik** [comment]
Единицей отображения в zencad - интерактивный объект (interactive_object).
Функция disp создаёт интерактивный объект и связывает его с дефолтной сценой.

```
#!/usr/bin/env python3
#coding: utf-8

from zencad import *

m=box(10)

interactive0 = disp(m)
interactive1 = disp(m)

interactive0.relocate(up(20))
interactive1.relocate(rotateZ(deg(45)))

show()
```

zencad.assemble.unit - это по сути сложная обёртка над интерактивными объектами. Она позволяет их стыковать между собой, перемещать группами или относительно друг друга, но логика над ними та же самое. Один юнит, одно отображение на сцене.

По exttrans. У вас extrans - это функция, возвращающая объект множественного преобразования, но не выполняющая его. Поэтому правильный синтаксис:
```
extrans()(konchevik)
```

**Comment by mirmik** [comment]
Мне нужно капитально обновить мануал на эту тему...

**Comment by oleg-medovikov** [comment]
```
def extrans(): return multitransform([
    translate(0,-rad-45,27) * rotateZ(deg(-90)) * rotateY(deg(165)),
    translate(0,-rad,-70),
    nulltrans()
])

extrans()(konchevik)
```

у меня так не получается

**Comment by mirmik** [comment]
А кочевник это unit?  Увы, юниты и интерактивные объекты пока перемещаются только через relocate. Дать им для совместимости также возможности, что и shape-ам в планах.

**Comment by oleg-medovikov** [comment]
```
kon1 = konchevik
kon2 = konchevik
kon1.relocate( translate(0,-rad-45,27) * rotateZ(deg(-90)) * rotateY(deg(165)))
kon2.relocate( translate(0,-rad+20,-40))
disp(kon2)
disp(kon1)
```
Ну вот делаю вот так, а отображается только последний.

**Comment by mirmik** [comment]
У вас kon1 и kon2 - это один и тот же объект. Вы его перенесли  сначала на одну позицию, потом на другую, потом дважды вызвали для него disp. Поскольку disp видит, что это уже объект отображения, он первый раз биндит его в сцену, а во второй раз вообще ничего не делает.

**Comment by mirmik** [comment]
Увы, но придётся создавать второго кочевника.

**Comment by mirmik** [comment]
```def make_kochevnik():
    u = zencad.assemble.unit()
    u.add_shape(shp0)
    u.add_shape(shp1, color=color.red)
    return u

Koch0=make_kochevnik()
koch1=make_kochevnik()
```

**Comment by oleg-medovikov** [comment]
Это не кочевник, а концевик. Просто требовалось  разместить 6 штук в проекте и метод multitrans выглядит очень симпатично, с точки зрения кода. В одном месте описаны координаты каждого.
 

![screenshot-2020-03-01-21:54:13](https://user-images.githubusercontent.com/57829409/75631831-4a578e80-5c07-11ea-8310-4ad172709e4b.png)

**Comment by mirmik** [comment]
Мысль понятна. Возможность двигать юниты я уже сделал для следующей версии. Добавить мультитрансу возможность их копировать сделать скорее всего будет несложно. 

Через какое-то время станет работать ожидаемым образом.
